packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -946,21 +984,51 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
 
   void _cursorTick(Timer timer) {
     _showCursor.value = !_showCursor.value;
+    if (widget.cursorOpacityAnimates) {
+      // If we want to show the cursor, we will animate the opacity to the value
+      // of 1.0, and likewise if we want to make it disappear, to 0.0. An easing
+      // curve is used for the animation to mimic the aesthetics of the native
+      // iOS cursor.
+      final double toValue = _showCursor.value ? 1.0 : 0.0;
+      _cursorController.animateTo(toValue, duration: _fadeDuration, curve: Curves.easeOut);
+    } else {
+      _cursorController.value = _showCursor.value ? 1.0 : 0.0;
+    }
+
     if (_obscureShowCharTicksPending > 0) {
-      setState(() { _obscureShowCharTicksPending--; });
+      setState(() {
+        _obscureShowCharTicksPending--;
+      });
     }
   }
 
+  void _cursorWaitForStart(Timer timer) {
+    _cursorTimer?.cancel();
+    _cursorTimer = Timer.periodic(_kCursorBlinkHalfPeriod, _cursorTick);
+  }
+
   void _startCursorTimer() {
     _showCursor.value = true;
-    _cursorTimer = Timer.periodic(_kCursorBlinkHalfPeriod, _cursorTick);
+    if (widget.cursorOpacityAnimates) {
+      _cursorController.value = 1.0;
+      _cursorController.addListener(_onCursorColorTick);
```

Comment:
Not sure if you need to keep adding/removing the listener. Since it never really changes and it's the same controller. You just need to stop and start the controller which you're already doing.

Reply:
Do I ever need to remove the listener?

Reply:
When you dispose